lightningd/dev_newhtlc.c

```diff
@@ -0,0 +1,125 @@
+#include <ccan/str/hex/hex.h>
+#include <daemon/jsonrpc.h>
+#include <daemon/log.h>
+#include <daemon/sphinx.h>
+#include <lightningd/channel/gen_channel_wire.h>
+#include <lightningd/lightningd.h>
+#include <lightningd/peer_control.h>
+#include <lightningd/subd.h>
+#include <utils.h>
+
+static bool offer_htlc_reply(struct subd *subd, const u8 *msg, const int *fds,
+			     struct command *cmd)
+{
+	u64 htlc_id;
+	u16 failcode;
+	u8 *failstr;
+
+	if (!fromwire_channel_offer_htlc_reply(msg, msg, NULL,
+					       &htlc_id, &failcode, &failstr)) {
+		command_fail(cmd, "Invalid reply from daemon: %s",
+			     tal_hex(msg, msg));
+		return true;
+	}
+
+	if (failcode != 0) {
+		command_fail(cmd, "failure %u: %.*s", failcode,
+			     (int)tal_len(failstr), (char *)failstr);
+	} else {
+		struct json_result *response = new_json_result(cmd);
+
+		json_object_start(response, NULL);
+		json_add_u64(response, "id", htlc_id);
+		json_object_end(response);
+		command_success(cmd, response);
+	}
+	return true;
+}
+
+static void json_dev_newhtlc(struct command *cmd,
+			     const char *buffer, const jsmntok_t *params)
+{
+	struct lightningd *ld = ld_from_dstate(cmd->dstate);
+	struct peer *peer;
+	u8 *msg;
+	jsmntok_t *peeridtok, *msatoshitok, *expirytok, *rhashtok;
+	unsigned int expiry;
+	u64 msatoshi;
+	struct sha256 rhash;
+	struct hoppayload *hoppayloads;
+	u8 sessionkey[32];
+	struct onionpacket *packet;
+	u8 *onion;
+	struct pubkey *path = tal_arrz(cmd, struct pubkey, 1);
+
+	if (!json_get_params(buffer, params,
+			     "peerid", &peeridtok,
+			     "msatoshi", &msatoshitok,
+			     "expiry", &expirytok,
+			     "rhash", &rhashtok,
+			     NULL)) {
+		command_fail(cmd, "Need peerid, msatoshi, expiry and rhash");
+		return;
+	}
+
+	peer = peer_from_json(ld, buffer, peeridtok);
+	if (!peer) {
+		command_fail(cmd, "Could not find peer with that peerid");
+		return;
+	}
+
+	if (peer->owner && !streq(peer->owner->name, "lightningd_channel")) {
+		command_fail(cmd, "Peer not in lightningd_channel (%s instead)",
+			     peer->owner ? peer->owner->name : "unattached");
+		return;
+	}
+
+	if (!streq(peer->condition, "Normal operation")) {
```

Comment:
Definitely.  We should have a "ready" flag....  I've added a FIXME where we do this.